src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.cs

```diff
@@ -14,6 +18,69 @@ internal static void VerifyValidPath(string path, string argName)
             }
         }
 
+        internal static bool Copy(string sourcePath, string destinationPath, bool recursive,
+            bool skipExistingFiles, CancellationToken cancellationToken)
+        {
+            bool success = true;
+            if (!DirectoryExists(sourcePath))
+            {
+                throw new DirectoryNotFoundException(sourcePath);
+            }
+
+            // Create destination directory if not exists
+            if (!DirectoryExists(destinationPath))
+            {
+                CreateDirectory(destinationPath);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            var fse = new FileSystemEnumerable<(string childPath, bool isDirectory)>(sourcePath,
+                static (ref FileSystemEntry entry) => (entry.ToFullPath(), entry.IsDirectory),
+                recursive ? EnumerationOptions.CompatibleRecursive : EnumerationOptions.Compatible);
+
+            foreach ((string childPath, bool isDirectory) in fse)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                try
+                {
+                    string basePath = childPath[sourcePath.Length..];
+                    if (PathInternal.IsDirectorySeparator(basePath[0]))
+                    {
+                        basePath = basePath[1..];
+                    }
+
+                    string destFilePath = Path.Join(destinationPath, basePath);
+                    if (isDirectory && !DirectoryExists(destFilePath))
```

Comment:
You can avoid the cost of `DirectoryExists` and call `CreateDirectory` directly.